miniconf/src/node.rs

```diff
@@ -0,0 +1,263 @@
+use core::{
+    fmt::Write,
+    iter::{Copied, Skip},
+    ops::{Deref, DerefMut},
+    slice::Iter,
+    str::Split,
+};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{Error, IntoKeys, KeysIter, Traversal, TreeKey};
+
+/// Type of a node: leaf or internal
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub enum NodeType {
+    /// A leaf node
+    ///
+    /// There are no nodes below this node
+    Leaf,
+
+    /// An internal node
+    ///
+    /// A non-leaf node with zero or more leaf nodes below this node
+    Internal,
+}
+
+impl NodeType {
+    /// Whether this node is a leaf
+    #[inline]
+    pub const fn is_leaf(&self) -> bool {
+        matches!(self, Self::Leaf)
+    }
+}
+
+/// Type and depth of a node in a `TreeKey`
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Node {
+    /// The node depth
+    ///
+    /// This is the length of the key required to identify it.
+    pub depth: usize,
+    /// Leaf or internal
+    pub typ: NodeType,
+}
+
+impl Node {
+    /// The depth
+    #[inline]
+    pub const fn depth(&self) -> usize {
+        self.depth
+    }
+
+    /// The node type
+    #[inline]
+    pub const fn typ(&self) -> NodeType {
+        self.typ
+    }
+
+    /// The node is a leaf node
+    #[inline]
+    pub const fn is_leaf(&self) -> bool {
+        self.typ.is_leaf()
+    }
+
+    /// Create a leaf node
+    #[inline]
+    pub const fn leaf(depth: usize) -> Self {
+        Self {
+            depth,
+            typ: NodeType::Leaf,
+        }
+    }
+
+    /// Create an inernal node
+    #[inline]
+    pub const fn internal(depth: usize) -> Self {
+        Self {
+            depth,
+            typ: NodeType::Internal,
+        }
+    }
+}
+
+impl From<Node> for usize {
+    fn from(value: Node) -> Self {
+        value.depth
+    }
+}
+
+/// Map a `TreeKey::traverse_by_key()` `Result` to a `NodeLookup::lookup()` `Result`.
+impl TryFrom<Result<usize, Error<()>>> for Node {
+    type Error = Traversal;
+    fn try_from(value: Result<usize, Error<()>>) -> Result<Self, Traversal> {
+        match value {
+            Ok(depth) => Ok(Node::leaf(depth)),
+            Err(Error::Traversal(Traversal::TooShort(depth))) => Ok(Node::internal(depth)),
+            Err(Error::Inner(depth, ())) => Err(Traversal::TooShort(depth)),
+            Err(Error::Traversal(err)) => Err(err),
+            Err(Error::Finalization(())) => unreachable!(),
+        }
+    }
+}
+
+/// Look up an `IntoKeys` in a `TreeKey` and transcode into `Self`.
+pub trait Transcode {
+    /// Perform a lookup and transcode the keys into self
```

Comment:
```suggestion
/// Look up an `IntoKeys` in a `TreeKey` and transcode into a `Node`.
pub trait Transcode {
    /// Perform a lookup and transcode the keys into a node
```

Reply:
I was confused reading these docs about what is being transcoded. It makes sense after the review that we're transcoding into other key types - we should specifically mention that instead of an ambiguous `Self` type of the trait implementer

Reply:
#218